### PR TITLE
Consolidated system observables

### PR DESCRIPTION
**Pull request by ernstsson**

### CODE REVIEW

**Comment by ernstsson** [comment]
@maxtroy @smile616
First step to flatten the calls more sensibly, and without the additional allocations the other flattener had to use.
Based on this I think I can add immediate calls to `observableActivate` and `update` when on the same looper. Have two more commits for those later

File: agera/src/main/java/com/google/android/agera/BaseObservable.java [hunk 1]
```diff
@@ -38,149 +39,131 @@
...
+      } else {
+        handler.obtainMessage(WorkerHandler.MSG_FIRST_ADDED, this).sendToTarget();
+      }
+    }
   }
 
   @Override
-  public final void removeUpdatable(@NonNull final Updatable updatable) {
+  public synchronized final void removeUpdatable(@NonNull final Updatable updatable) {
     checkState(Looper.myLooper() != null, "Can only be removed on a Looper thread");
```
Comment (smile616) [inline agera/src/main/java/com/google/android/agera/BaseObservable.java]: "Can only be removed on a Looper thread" is it actually true?

File: agera/src/main/java/com/google/android/agera/BaseObservable.java [hunk 2]
```diff
@@ -38,149 +39,131 @@
...
-    worker.removeUpdatable(updatable);
+    remove(updatable);
+    if (size == 0) {
+      handler.obtainMessage(MSG_LAST_REMOVED, this).sendToTarget();
+      handler.removeMessages(MSG_UPDATE, this);
```
Comment (smile616) [inline agera/src/main/java/com/google/android/agera/BaseObservable.java]: maybe remove before send new message?

File: agera/src/main/java/com/google/android/agera/BaseObservable.java [hunk 3]
```diff
@@ -38,149 +39,131 @@
...
-  private final Worker worker;
+  private static final Object[] NO_UPDATABLES_OR_HANDLERS = new Object[0];
+  @NonNull
+  private final WorkerHandler handler;
+  private final int shortestUpdateWindowMillis;
+
+  @NonNull
+  private Object[] updatablesAndHandlers;
+  private int size;
+  private long lastUpdateTimestamp;
```
Comment (smile616) [inline agera/src/main/java/com/google/android/agera/BaseObservable.java]: logically `lastUpdateTimestamp` should be next to `shortestUpdateWindowMillis`

File: agera/src/main/java/com/google/android/agera/CompiledRepository.java [hunk 4]
```diff
@@ -49,10 +49,9 @@ static Repository compiledRepository(
       @NonNull final Merger<Object, Object, Boolean> notifyChecker,
       @RepositoryConfig final int concurrentUpdateConfig,
       @RepositoryConfig final int deactivationConfig) {
-    final Observable eventSource = perMillisecondObservable(frequency,
-        compositeObservable(eventSources.toArray(new Observable[eventSources.size()])));
     final Object[] directiveArray = directives.toArray();
```
Comment (smile616) [inline agera/src/main/java/com/google/android/agera/CompiledRepository.java]: new Object[derectives.size()]

File: agera/src/main/java/com/google/android/agera/Observables.java [hunk 5]
```diff
@@ -45,17 +43,23 @@
...
   public static Observable compositeObservable(@NonNull final Observable... observables) {
+    return compositeObservable(0, observables);
+  }
+
+  @NonNull
+  static Observable compositeObservable(final int shortestUpdateWindowMillis,
+      @NonNull final Observable... observables) {
     if (observables.length == 0) {
-      return new CompositeObservable();
+      return new CompositeObservable(0);
```
Comment (smile616) [inline agera/src/main/java/com/google/android/agera/Observables.java]: do we have NullObservable or something like that?

File: agera/src/main/java/com/google/android/agera/BaseObservable.java (hunk 3 above)
Comment (ernstsson) [inline agera/src/main/java/com/google/android/agera/BaseObservable.java]: we're grouping by final/non-final in the rest of the codebase though, so not really possible to group by use case as such (and follow the current style)

File: agera/src/main/java/com/google/android/agera/BaseObservable.java (hunk 2 above)
Comment (ernstsson) [inline agera/src/main/java/com/google/android/agera/BaseObservable.java]: this is just an inline of an existing method, would rather keep it as is in this commit, and iterate on that later

File: agera/src/main/java/com/google/android/agera/BaseObservable.java (hunk 1 above)
Comment (ernstsson) [inline agera/src/main/java/com/google/android/agera/BaseObservable.java]: Same here, just moved but; yes, it's actually true. It keeps the references in a `ThreadLocal`, add/remove needs to be called paired from the same thread. This is also described in the javadoc for `Observable`

File: agera/src/main/java/com/google/android/agera/CompiledRepository.java (hunk 4 above)
Comment (ernstsson) [inline agera/src/main/java/com/google/android/agera/CompiledRepository.java]: Hm, perhaps, but lets keep it out of this PR and ping @maxtroy about it

File: agera/src/main/java/com/google/android/agera/Observables.java (hunk 5 above)
Comment (ernstsson) [inline agera/src/main/java/com/google/android/agera/Observables.java]: `CompositeObservable` doubles as a null observable. Even if it's not doing anything it still needs to follow the contract of throwing exceptions for non paired calls etc (purely a safety precaution) so in the end an empty `CompositeObservable` will look exactly the same

File: agera/src/main/java/com/google/android/agera/BaseObservable.java (hunk 3 above)
Comment (smile616) [inline agera/src/main/java/com/google/android/agera/BaseObservable.java]: maybe we should consider change in the style?

**Comment by smile616** [comment]
LGTM

File: agera/src/main/java/com/google/android/agera/BaseObservable.java [hunk 6]
```diff
@@ -38,149 +39,131 @@
...
-    worker = new Worker(this);
+    this.shortestUpdateWindowMillis = shortestUpdateWindowMillis;
+    this.handler = workerHandler();
+    this.updatablesAndHandlers = NO_UPDATABLES_OR_HANDLERS;
+    this.size = 0;
   }
 
   @Override
-  public final void addUpdatable(@NonNull final Updatable updatable) {
+  public synchronized final void addUpdatable(@NonNull final Updatable updatable) {
```
Comment (maxtroy) [inline agera/src/main/java/com/google/android/agera/BaseObservable.java]: I would prefer a private object for locking (`synchronized` on). That was one (implicit) responsibility of the Worker instance before this commit.

File: agera/src/main/java/com/google/android/agera/BaseObservable.java [hunk 7]
```diff
@@ -38,149 +39,131 @@
...
-      throw new IllegalStateException("Updatable not added, cannot remove.");
     }
+    throw new IllegalStateException("Updatable not added, cannot remove.");
+  }
 
-    synchronized void sendUpdate() {
+  synchronized void sendUpdate() {
+    handler.removeMessages(WorkerHandler.MSG_UPDATE, this);
+    final long elapsedRealtimeMillis =
+        shortestUpdateWindowMillis > 0 ? elapsedRealtime() : 0;
```
Comment (maxtroy) [inline agera/src/main/java/com/google/android/agera/BaseObservable.java]: Despite the logical correctness involving `shortestUpdateWindowMillis <= 0`, I would prefer elapsedRealtimeMillis and lastUpdateTimestamp updated as usual, so I don't run into difficulty reading the following line:

```
timeFromLastUpdate = elapsedRealtimeMillis - lastUpdateTimestamp
                   = 0 - some timestamp
                   = negative result
```

Or, if your concern is about calling elapsedRealtime() unnecessarily, then a short-cutting if-block is better:

```
if (shortestUpdateWindowMillis > 0) {
  final long elapsedRealtimeMillis = elapsedRealtime();
  final long timeFromLastUpdate = elapsedRealtimeMillis - lastUpdateTimestamp;
  if (timeFromLastUpdate  < shortestUpdateWindowMillis) {
    handler.sendMessageDelayed(handler.obtainMessage(WorkerHandler.MSG_UPDATE, this),
        shortestUpdateWindowMillis - timeFromLastUpdate);
    return;
  }
  lastUpdateTimestamp = elapsedRealtimeMillis;
}
// dispatch updates
for (int index = 0; ...
```

**Pull request merged**